File: dotnet/tests/Pactman.NonprofitCheckPlus.Tests/Support/Fixtures.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text.Json.Nodes;
using Pactman.NonprofitCheckPlus.Configuration;

namespace Pactman.NonprofitCheckPlus.Tests.Support;

/// <summary>Shared fixtures for the test suite.</summary>
public static class Fixtures
{
    /// <summary>A key that must never appear in any diagnostic output.</summary>
    public const string ApiKey = "pactman_test_key_do_not_leak_8f2b";

    public const string BaseUrl = "http://mock.test";

    /// <summary>A client wired to <paramref name="handler"/>, using the documented setup.</summary>
    public static PactmanClient Client(
        FakeHandler handler,
        Clock? clock = null,
        TimeSpan? timeout = null,
        RetryOptions? retry = null,
        double? maxRequestsPerSecond = null,
        IReadOnlyDictionary<string, string>? defaultHeaders = null)
    {
        var options = new PactmanClientOptions
        {
            ApiKey = ApiKey,
            BaseUrl = BaseUrl,
            Timeout = timeout,
            Retry = retry,
            MaxRequestsPerSecond = maxRequestsPerSecond,
            HttpClient = new HttpClient(handler) { Timeout = System.Threading.Timeout.InfiniteTimeSpan },
            DisposeHttpClient = true,
        };

        if (defaultHeaders is not null)
        {
            foreach (var header in defaultHeaders)
            {
                options.DefaultHeaders[header.Key] = header.Value;
            }
        }

        return new PactmanClient(options, clock?.Hooks());
    }

    /// <summary>A representative organization, mirroring the published response example.</summary>
    public static JsonObject Nonprofit(Action<JsonObject>? customize = null)
    {
        var organization = new JsonObject
        {
            ["pactman_org_url"] = "https://pactman.org/profile/nonprofit/example-nonprofit-r5U9r8yRcZ",
            ["organization_info_last_modified"] = "2/22/2026 1:16:30 AM",
            ["ein"] = "411787097",
            ["organization_name"] = "EXAMPLE NONPROFIT",
            ["organization_name_aka"] = "EXAMPLE N.P",
            ["address_line1"] = "50 LOWELL AVE",
            ["address_line2"] = "APT 3B",
            ["city"] = "WESTFIELD",
            ["state"] = "MA",
            ["state_name"] = "Massachusetts",
            ["zip"] = "01085-2643",
            ["filing_req_code"] = "00",
            ["pub78_church_message"] = null,
            ["pub78_organization_name"] = "Example Nonprofit",
            ["pub78_ein"] = "411787097",
            ["pub78_verified"] = true,
            ["pub78_city"] = "Westfield",
            ["pub78_state"] = "MA",
            ["pub78_indicator"] = "0",
            ["organization_types"] = new JsonArray
            {
                new JsonObject
                {
                    ["organization_type"] = "Deductions for donations to public charities are generally limited...",
                    ["deductibility_limitation"] = "50%",
                    ["deductibility_status_description"] = "PC",
                },
            },
            ["most_recent_pub78"] = "12/12/2025 12:00:00 AM",
            ["bmf_church_message"] = null,
            ["bmf_organization_name"] = "EXAMPLE NONPROFIT",
            ["bmf_ein"] = "411787097",
            ["bmf_status"] = true,
            ["most_recent_bmf"] = "12/09/2025 12:00:00 AM",
            ["bmf_subsection"] = "03",
            ["subsection_description"] = "501(c)(3) Public Charity",
            ["foundation_code"] = "10",
            ["foundation_code_description"] = "Public charity described in section 509(a)(1) or (2)",
            ["ruling_month"] = "07",
            ["ruling_year"] = "2024",
            ["group_exemption"] = "0000",
            ["exempt_status_code"] = "01",
            ["ofac_status"] = "This organization was NOT included in the Office of Foreign Assets "
                + "Control Specially Designated Nationals (SDN) list.",
            ["revocation_code"] = null,
            ["revocation_date"] = null,
            ["reinstatement_date"] = null,
            ["irs_bmf_pub78_conflict"] = false,
            ["foundation_509a_status"] = "N/A",
            ["report_date"] = "3/25/2026 3:28:54 PM",
            ["foundation_type_code"] = "pc",
            ["foundation_type_description"] = "Public charity described in section 509(a)(1) or (2)",
        };

        customize?.Invoke(organization);

        return organization;
    }

    /// <summary>Wraps data in the envelope the API returns.</summary>
    public static JsonObject Envelope(JsonNode? data, Action<JsonObject>? customize = null)
    {
        var envelope = new JsonObject
        {
            ["code"] = 200,
            ["message"] = "OK",
            ["errors"] = null,
            ["data"] = data,
            ["timeTaken"] = 3,
            ["nonprofit_check_count"] = 1,
        };

        customize?.Invoke(envelope);

        return envelope;
    }

    /// <summary>An organization list, as the bulk endpoint returns it.</summary>
    public static JsonArray Organizations(params JsonNode[] organizations)
    {
        var list = new JsonArray();

        foreach (var organization in organizations)
        {
            list.Add(organization);
        }

        return list;
    }
}

File: java/sdk/src/test/java/org/pactman/nonprofitcheckplus/support/Fixtures.java
package org.pactman.nonprofitcheckplus.support;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.Map;
import org.pactman.nonprofitcheckplus.PactmanClient;
import org.pactman.nonprofitcheckplus.config.PactmanClientOptions;
import org.pactman.nonprofitcheckplus.http.TransportHooks;
import org.pactman.nonprofitcheckplus.internal.Json;

/** Shared test data and client wiring. */
public final class Fixtures {

    /** A key that must never appear in any diagnostic output. */
    public static final String TEST_API_KEY = "pactman_test_key_do_not_leak_8f2b";

    /** The base URL the stubbed clients point at. */
    public static final String BASE_URL = "http://mock.test";

    private Fixtures() {
    }

    /**
     * A client wired to a stub exchange.
     *
     * @param exchange the exchange to send through.
     * @return the client.
     */
    public static PactmanClient client(StubExchange exchange) {
        return client(exchange, null);
    }

    /**
     * A client wired to a stub exchange and a recording clock.
     *
     * @param exchange the exchange to send through.
     * @param hooks    the clock and random source, or {@code null} for the real ones.
     * @return the client.
     */
    public static PactmanClient client(StubExchange exchange, TransportHooks hooks) {
        return new PactmanClient(
                PactmanClientOptions.builder()
                        .apiKey(TEST_API_KEY)
                        .baseUrl(BASE_URL)
                        .httpExchange(exchange)
                        .build(),
                hooks);
    }

    /**
     * A representative organization, mirroring the published response example.
     *
     * @return the organization's fields, mutable so a test can vary one.
     */
    public static Map<String, Object> nonprofit() {
        Map<String, Object> organization = new LinkedHashMap<>();
        organization.put(
                "pactman_org_url",
                "https://pactman.org/profile/nonprofit/example-nonprofit-r5U9r8yRcZ");
        organization.put("organization_info_last_modified", "2/22/2026 1:16:30 AM");
        organization.put("ein", "411787097");
        organization.put("organization_name", "EXAMPLE NONPROFIT");
        organization.put("organization_name_aka", "EXAMPLE N.P");
        organization.put("address_line1", "50 LOWELL AVE");
        organization.put("address_line2", "APT 3B");
        organization.put("city", "WESTFIELD");
        organization.put("state", "MA");
        organization.put("state_name", "Massachusetts");
        organization.put("zip", "01085-2643");
        organization.put("filing_req_code", "00");
        organization.put("pub78_church_message", null);
        organization.put("pub78_organization_name", "Example Nonprofit");
        organization.put("pub78_ein", "411787097");
        organization.put("pub78_verified", Boolean.TRUE);
        organization.put("pub78_city", "Westfield");
        organization.put("pub78_state", "MA");
        organization.put("pub78_indicator", "0");

        Map<String, Object> organizationType = new LinkedHashMap<>();
        organizationType.put(
                "organization_type",
                "Deductions for donations to public charities are generally limited...");
        organizationType.put("deductibility_limitation", "50%");
        organizationType.put("deductibility_status_description", "PC");
        organization.put("organization_types", new ArrayList<>(Arrays.asList(organizationType)));

        organization.put("most_recent_pub78", "12/12/2025 12:00:00 AM");
        organization.put("bmf_church_message", null);
        organization.put("bmf_organization_name", "EXAMPLE NONPROFIT");
        organization.put("bmf_ein", "411787097");
        organization.put("bmf_status", Boolean.TRUE);
        organization.put("most_recent_bmf", "12/09/2025 12:00:00 AM");
        organization.put("bmf_subsection", "03");
        organization.put("subsection_description", "501(c)(3) Public Charity");
        organization.put("foundation_code", "10");
        organization.put(
                "foundation_code_description",
                "Public charity described in section 509(a)(1) or (2)");
        organization.put("ruling_month", "07");
        organization.put("ruling_year", "2024");
        organization.put("group_exemption", "0000");
        organization.put("exempt_status_code", "01");
        organization.put(
                "ofac_status",
                "This organization was NOT included in the Office of Foreign Assets Control "
                        + "Specially Designated Nationals (SDN) list.");
        organization.put("revocation_code", null);
        organization.put("revocation_date", null);
        organization.put("reinstatement_date", null);
        organization.put("irs_bmf_pub78_conflict", Boolean.FALSE);
        organization.put("foundation_509a_status", "N/A");
        organization.put("report_date", "3/25/2026 3:28:54 PM");
        organization.put("foundation_type_code", "pc");
        organization.put(
                "foundation_type_description",
                "Public charity described in section 509(a)(1) or (2)");

        return organization;
    }

    /**
     * The representative organization with fields replaced.
     *
     * @param overrides field names and values to set, in pairs.
     * @return the organization's fields.
     */
    public static Map<String, Object> nonprofit(Object... overrides) {
        Map<String, Object> organization = nonprofit();

        for (int i = 0; i + 1 < overrides.length; i += 2) {
            organization.put((String) overrides[i], overrides[i + 1]);
        }

        return organization;
    }

    /**
     * Wraps a payload in the envelope the API returns.
     *
     * @param data      the payload.
     * @param overrides envelope fields to replace, in pairs.
     * @return the envelope's fields.
     */
    public static Map<String, Object> envelope(Object data, Object... overrides) {
        Map<String, Object> envelope = new LinkedHashMap<>();
        envelope.put("code", 200L);
        envelope.put("message", "OK");
        envelope.put("errors", null);
        envelope.put("data", data);
        envelope.put("timeTaken", 3L);
        envelope.put("nonprofit_check_count", 1L);

        for (int i = 0; i + 1 < overrides.length; i += 2) {
            envelope.put((String) overrides[i], overrides[i + 1]);
        }

        return envelope;
    }

    /**
     * Serializes a value as JSON.
     *
     * @param value the value.
     * @return the JSON text.
     */
    public static String json(Object value) {
        return Json.write(value);
    }

    /**
     * Parses JSON.
     *
     * @param text the JSON text.
     * @return the decoded value.
     */
    public static Object parse(String text) {
        return Json.parse(text);
    }
}
